src/shared/tests.h

```diff
@@ -201,41 +201,62 @@ static inline int run_test_table(void) {
 #define DEFINE_TEST_MAIN(log_level)                     \
         DEFINE_TEST_MAIN_FULL(log_level, NULL, NULL)
 
-#define ASSERT_OK(expr)                                                                                  \
-        ({                                                                                               \
-                int _result = (expr);                                                                    \
-                if (_result < 0) {                                                                       \
-                        log_error_errno("Assertion failed: %s (result: %d, error: %m)", #expr, _result); \
-                        abort();                                                                         \
-                }                                                                                        \
+#define FMT(_pre_, x, sign) _Generic((x),                 \
```

Comment:
We should do this differently. The FMT() should take a single argument and then you can use it in the format string itself.

The usage should look like this:

log_error("%s:%i: Assertion failed: expected %s == %s, but " FMT(_expr1) " != " FMT(_expr2)

Reply:
This also needs to handle the bool type

Reply:
Thanks for this @DaanDeMeyer , when I tried to change it to single argument, I am getting some errors. 

```
In file included from ../src/basic/macro.h:386,
                 from ../src/basic/architecture.h:6,
                 from ../src/test/test-gpt.c:3:
../src/shared/tests.h:236:35: error: format ‘%s’ expects a matching ‘char *’ argument [-Werror=format=]
  236 |                         log_error("%s:%i: Assertion failed: expected %s == %s, but " FMT(_expr1) " != " FMT(_expr2), \
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/basic/log.h:214:86: note: in definition of macro ‘log_full_errno_zerook’
  214 |                         ? log_internal(_level, _e, PROJECT_FILE, __LINE__, __func__, __VA_ARGS__) \
      |                                                                                      ^~~~~~~~~~~
../src/basic/log.h:246:28: note: in expansion of macro ‘log_full’
  246 | #define log_error(...)     log_full(LOG_ERR,     __VA_ARGS__)
      |                            ^~~~~~~~
../src/shared/tests.h:236:25: note: in expansion of macro ‘log_error’
  236 |                         log_error("%s:%i: Assertion failed: expected %s == %s, but " FMT(_expr1) " != " FMT(_expr2), \
      |                         ^~~~~~~~~
../src/test/test-gpt.c:106:9: note: in expansion of macro ‘ASSERT_EQ’
  106 |         ASSERT_EQ(x.designator, y.designator);
      |         ^~~~~~~~~
In file included from ../src/test/test-gpt.c:10:
../src/shared/tests.h:236:77: note: format string is defined here
  236 |                         log_error("%s:%i: Assertion failed: expected %s == %s, but " FMT(_expr1) " != " FMT(_expr2), \
      |                                                                            ~^
      |                                                                             |
      |                                                                             char *

```

some part of the error. kindly help on this, thank you.

Reply:
@Unique-Usman Please push your latest changes, otherwise I can't help out

Reply:
@DaanDeMeyer noted, should I push the other changes in the review ?

Reply:
@Unique-Usman I have just pushed, please do `git pull --rebase`, maybe fix the merge conflicts, and then push again.

Reply:
@DaanDeMeyer I have just pushed, I added the bool data type which you said earlier, I also move the DECIMAL_STR_FMT to the macro.h. I also added the changes by @yuwata